Bugfix: multiple configuration files not processed

rbailey-godaddy opened the pull request:
To help us get this pull request reviewed and merged quickly, please be sure to include the following items:

* [X] Tests (if applicable)
* [X] Documentation (if applicable)
* [X] Changelog entry
* [X] A full explanation here in the PR description of the work done

## PR Type
What kind of change does this PR introduce?

* [X] Bugfix
* [ ] Feature
* [ ] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [ ] Build related changes
* [ ] CI related changes
* [ ] Documentation content changes
* [X] Tests
* [ ] Other

## Backward Compatibility

Is this change backward compatible with the most recently released version? Does it introduce changes which might change the user experience in any way? Does it alter the API in any way?

* [X] Yes (backward compatible)
* [ ] No (breaking changes)

This is arguable but I believe "backward compatible" best describes the spirit of the change.

Behavior has changed, but the new behavior matches what users thought was the behavior all along (i.e. if you specify multiple configuration files, all specified configuration files will be used). The `read_project_toml()` function still returns a string, but in the event multiple files were specified, that string is a comma-delimited list of filenames instead of a single filename. However, no in-tree code ever looks at this return value (except for some unit tests).

## Issue Linking
<!--
    KEYWORD #ISSUE-NUMBER
    [closes|fixes|resolves] #
-->
fixes #467 

## What's new?
Several configuration file processing bugs have been fixed.

* When the `--config` parameter appears multiple times, tartufo now processes each specified file in order; previously it ignored all but the last file. The documentation was silent on whether this was a valid thing or not, but no warning or error was generated and users were doing this and believing it would do something useful.
* When running `scan-folder`, an oversight caused the target folder not to be scanned for configuration files. Now tartufo will scan for configuration files, the same way it does for `scan-local-repo` and `scan-remote-repo`
* The same problem occurred for `pre-commit`.
* The logic that was supposed to avoid reloading a configuration file was horribly broken before, and it was even more broken by added support for multiple `--config` options. This has been fixed.
* To reduce the chances of screwing this up in the future, Configuration processing has been pulled into a `load_config()` method and all of the open-coded stuff -- typically in `load_repo()` although it had nothing to do with repository loading -- has been replaced with calls to the new method

Additionally, merging of data from multiple files is improved. (This wasn't really a factor before, because tartufo never looked at more than one user-specified configuration file.) Single-valued attributes are overwritten (so the last configuration file that specifies a setting wins) and list-valued attributes are concatenated (generally aligning to the expected behavior). The documentation has been updated to describe this, and unit tests have been added to verify that's really what the code does.

The documentation updates also include greatly expanded material on how configuration files are detected, since that appears to be a point of enduring confusion among both users and maintainers. ;-)

rscottbailey committed:
Multiple --config handling fixups

Correct a number of defects with configuration file handling:

* We never said anything one way or the other, but users assumed
  `--config` could be specified multiple times on the command line.
  This didn't generate warnings or errors, but only the last option
  was processed. Mark this option with `multiple=True`
* Adjust `read_pyproject_toml()` to accept multiple config file inputs.
  * Put nearly entire function body into a loop on `value`, which now
    is a tuple instead of a string; early `return`s now are `continue`s
  * Function output, which used to be the name of the configuration
    file, now is a comma-delimited list of files (still a string);
    nothing actually uses this so it probably doesn't matter
  * When saving configuration to `ctx.default_map`, we concatenate
    list-valued members instead of just overwriting them
* Also in `read_pyproject_toml()`, `target_args` did not include
  `"target"` (used by `scan-folder`) and consequently did not discover
  configuration files in the target folder. This has been fixed and
  commented a little more liberally to clarify that usage.

Existing unit tests still pass (after adjusting to pass a tuple instead
of a string) -- new unit tests need to be written but I have verified
interactively that we now get the expected behavior.

rscottbailey committed:
Add unit tests for multiple configuration files

Add tests to verify the asserted behavior of parameter loading from
multiple configuration files

rscottbailey committed:
Add CHANGELOG entry

rscottbailey committed:
Documentation fixups

Explicitly describe the behavior when multiple configuration files are
specified, and clarify (somewhat) the search behavior

rscottbailey committed:
mypy linting fixups

The Good: use correct annotation for `value` parameter to indicate it is
a tuple consisting of an indeterminate number of strings, instead of a
tuple consisting of exactly one string.

The Bad: fix some unit tests where I forgot to convert the `value`
parameter into a tuple in the first place.

The Ugly: `read_pyproject_toml()` has been systematically abusing
click's context `default_map` by treating it like a dict even though it
actually is defined as an (immutable) `Mapping`. It's been doing this
for a long time, and evidently it works. A proper fix (probably using
`params` instead of `default_map`) is not obvious to me -- I tried just
switching to params, and it ate my updates. Therefore, we continue the
existing abusive practice, document the problem, and add an annotation
to silence the (valid) mypy error.

rscottbailey committed:
Placate pylint

Eliminate an unused variable from one of the new unit tests

sushantmimani committed:
Merge branch 'main' into bugfix/multi-config

rscottbailey committed:
Harmonize configuration file discovery

After some poking around, it is clear that the path to "local"
configuration files lies from the scanner instances, which call
`load_config_from_path()` directly, bypassing `read_pyproject_toml()`
entirely. The latter appears to be used only for files specified on the
command line with `--config`.

Folder scanning (using class `FolderScanner`) did not ever look for
configuration files. This class has no `load_repo()` method like the git
scanners, so clone the relevant logic directly into the `__init__()`
method. This appears after the parent init call, because it appears to
be most congruent to the path to `load_repo()`, which is defined in the
specific classes but called from the `ScannerBase` class init method.

Somewhat similarly, pre-commit scanning (using class
`GitPreCommitScanner`) did not ever look for configuration files. This
class has a `load_repo()` method, but it didn't contain the appropriate
logic. Clone logic into that, also.

rscottbailey committed:
Simplify read_pyproject_toml()

Now that we know this is called only in response to use of `--config`,
we can perform some logic reductions.

* Because `ctx.params` is always empty, the `target_args` search always
  fails and `config_path` always gets set to the current directory. Just
  do that initially before entering the loop and remove the distracting
  pointless tests.
* Because we are called only when the user specifies a file, the "fail
  silently on implicit search" logic never is triggered; remove it. We
  always raise an exception if the (exact as-specified) file is not
  found.
* Similarly, if we barf on a file containing garbage, we know the name
  of the file, so no name construction is required.

No change to unit tests except to eliminate a test that was evaluating
behavior for one of the "will never happen" conditions.

rscottbailey committed:
More read_pyproject_toml() simplification

Because we now know that `read_pyproject_toml()` only gets called once
and does not need to accumulate data across multiple calls, we can clean
up the context `default_map` usage.

* Assume map is empty (`None`) on entry
* Accumulate settings in a local `config_data` dict
* Just before exiting, stuff everything into `default_map` as a complete
  replacement. Neither pylint nor mypy complain about this

Replaces long "FIXME" comment about (immutable) mapping updates.

rscottbailey committed:
Reduce configuration duplicity

Why is every class doing almost identically the same thing to load
configuration? Bleah.

Add a `load_config()` method to the base `ScannerBase` class; this does
the configuration lookup once, in one place, one way.

For all of the derived classes, remove the open-coded crap (generally in
`load_repo()` even though it has nothing to do with loading the
repository) and instead make a call to `self.load_config()` from the end
of the `__init__()` method. Note that we can't simply delegate this to
the base class's initializer, because it needs to happen after the files
exist, which is after `load_repo()` in the case of remote repositories.

rscottbailey committed:
Fix config "overlap" brain-damage

It's not clear this really is a problem, but we had logic that
preventing reading a "local" config file if we'd already processed it
globally. However that logic was broken, and it got even more broken
when we starting allowing multiple global configuration files.

First, get rid of return value from `read_pyproject_toml()` -- which was
unrecoverable -- in favor of stashing a set of processed configuration
files globally. Make these fully-resolved Path objects to avoid any
ambiguity in what was or wasn't referenced.

While we're at it, junk the return value which provided nothing useful.
Also while we're at it, use this ourselves and avoid reusing a file if
the user says something stupid like `--config foo --config foo`.

I almost added a test to avoid re-reading a file in the case where the
user says `--config foo --config foo`, but I didn't do that so that the
scenario `--config a --config b --config a` preserves "last reference
wins" semantics, which wouldn't be the case if we short-circuited the
second read of a because we'd seen it already.

Second, replace the test against the global options -- which had the
wrong type, and which had what the user typed instead of the resolved
name, and test for membership in the processed set instead.

rscottbailey committed:
Beating a dead horse department

Significant extend and reword (again) the material on configuration
processing, hopefully clarifying how configuration discovery works and
how options from multiple files are merged.

rscottbailey committed:
pylint fixups

rscottbailey committed:
Fix target configuration loading

The call to `load_confg()` must occur *BEFORE* calling `load_repo()`.
That's because typically the latter routine will call
`filter_submodules()`, which as it hamhandedly generates new filter
expressions, causes processing and memoization of some of the
configuration settings (i.e. `exclude_path_patterns`) with the result
that confguration loads after this point basically get ignored.

Move up the configuration load to avoid this problem, and put a hateful
FIXME comment in the submodule kludging. I didn't do more because that
would have involved fixing a bunch of related unit tests. :-p

sushantmimani commented on docs/source/configuration.rst:
```diff
@@ -8,27 +8,98 @@ unwieldy and lead to an overly cumbersome command. It also becomes difficult to
+Therefore, while it normally a target's configuration file will override settings
+specified using ``--config`` on the command line, this behavior can be overridden
+by specifying the target's configuration explicitly first, and then the desired
+"master" configuration second.
```
I think this paragraph might be missing a few words.

rbailey-godaddy replied:
Actually, on re-read, I think it has a word or two too many :) -- let me try to clean up that and the one below

sushantmimani commented on docs/source/configuration.rst:
```diff
@@ -8,27 +8,98 @@ unwieldy and lead to an overly cumbersome command. It also becomes difficult to
+
+Note this behavior is not applicable to remote repository scans, because the
+remote repository will be cloned to a scratch directory and neither that directory
+nor the configuration files within it are available to specify with ``-config``.
```
We should probably make the `note` bold to make it stand out?

emayuri-godaddy reviewed (approved):
LGTM

rscottbailey committed:
Documentation cleanups

A little editorial work to fix problems pointed out by @sushantmimani
during PR review.

sushantmimani reviewed (approved):
🚀

jsimmons1-godaddy commented on docs/source/configuration.rst:
```diff
@@ -8,27 +8,98 @@ unwieldy and lead to an overly cumbersome command. It also becomes difficult to
+   tartufo --config tartufo.toml --config myconfig.toml scan-local-repo .
+
+will cause ``tartufo`` to read ``tartufo.toml`` in the current directory
+(assuming it exists, and in parent directories otherwise), and then ``myconfig.toml``
```
`(assuming it exists, and in parent directories otherwise)` is this accurate? Unless the behavior changed in this PR (which I don't see, but is quite possible), it will not traverse upwards in this case, because `tartufo.toml` gets turned into an absolute path to the file.

rbailey-godaddy replied:
I believe this is still correct (although we plan a PR before the next release that will unilaterally remove the traversal logic, which we believe is effectively unused and won't be missed).

If you pass a relative path via `--config`, that is sent to `config.read_pyproject_toml()`, which sends it (unmodified) to `config.load_config_from_path()` (with `config_path` set to `.` and  `traverse` implicitly `True`).

If it can't find a file, `load_config_from_path()` recursively calls itself, except this time `config_path` is set to the parent of the original path (i.e. `../`) and repeats this until you reach the root directory, at which point `config_path.parent` matches `config_path` and the loop terminates.

jsimmons1-godaddy replied:
Okay, thanks. Yeah it sound like it will become immaterial soon, but just for completion: my understanding of the current/old behavior is: If you pass a relative path via `--config` , the click library transforms it immediately into an absolute path. The rest appears correct, except it is effectively unable to traverse due to that transformation that has already been done.

rbailey-godaddy committed:
Merge branch 'main' into bugfix/multi-config
